### ISSUE TITLE
Malloc error 

### ISSUE THREAD

**Issue by theweasley** [quietvoid/hdr10plus_parser#20]
First off, this is a fantastic script. Been doing some HDR10+ metadata retention with it. Thank you!

Now on to the issue at hand.

The script works well for most titles with HDR10+ metadata but then throws malloc errors on others. I'm not sure if anyone else is having the same experience.

I encode on a headless Ubuntu 18.04. This is the error that I just recently got:
```x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main 10 profile, Level-5.1 (Main tier)
x265 [info]: Thread pool created using 8 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(17 rows)
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 3
x265 [info]: Keyframe min / max / scenecut / bias  : 23 / 250 / 40 / 5.00
x265 [info]: Lookahead / bframes / badapt        : 20 / 4 / 2
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 3 / off / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 3 / 1.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-16.0 / 0.60
x265 [info]: VBV/HRD buffer / max-rate / init    : 36000 / 18000 / 0.900
x265 [info]: tools: rd=3 psy-rd=1.00 rdoq=1 psy-rdoq=3.00 early-skip
x265 [info]: tools: rskip mode=1 signhide tmvp b-intra strong-intra-smoothing
x265 [info]: tools: lslices=6 deblock(tC=-3:B=-3) dhdr10-info
ffmpeg: malloc.c:3830: _int_malloc: Assertion `chunk_main_arena (bck->bk)' failed.speed=0.367x
```

**Comment by quietvoid** [comment]
Can you mention the title the metadata is from? I have heard issues about x265 not being able to complete encodes on some.

**Comment by theweasley** [comment]
The new Back to the Future UHD with HDR10+

**Comment by quietvoid** [comment]
> The new Back to the Future UHD with HDR10+
> 1917

At least for 1917, there are a lot of mentions the issue is with x265. Supposedly NVENC with the metadata has no problem encoding fully.
x265 is terminating a few frames before the end, and the JSON has the same number of metadata messages as the source frame number.

I'd have to look at Back to the Future but it is possible that the same problem happens, which doesn't seem to be related to the parser.

**Comment by theweasley** [comment]
Hmm interesting. But it seems those guys have been able to encode some frames. With me, the x265 encoder crashes with malloc errors within the first 1-2 seconds of encoding. Doesn't even start properly. It does seem to be a memory issue with x265 judging my those malloc errors. But then again for some other HDR10+ encodes, it has worked completely file. Figured I'd check in here also.

But you're right, it doesn't seem to be a problem with the parsing script itself. Guess our best bet is to wait for a fix with x265. I've compiled another ffmpeg build with x265 3.5 also. But that doesn't fix the issue either.

**Comment by quietvoid** [comment]
I'll keep the issue open until I can have a look at the new discs.

**Comment by theweasley** [comment]
> I'll keep the issue open until I can have a look at the new discs.

Thanks man. Most appreciated.

**Comment by theweasley** [comment]
Getting the same issue with Back to the Future Part II

**Comment by quietvoid** [comment]
x265 is definitely the one crashing, but I have an idea what is going wrong.
```
encoded 222 frames in 8.15s (27.23 fps), 371.76 kb/s, Avg QP:32.79  
free(): invalid next size (fast)`

**Comment by quietvoid** [comment]
@theweasley I've pushed a possible fix, can you test a build from master?

**Comment by theweasley** [comment]
@quietvoid okay giving it a go now. Will report back.

**Comment by theweasley** [comment]
@quietvoid Getting this error now with your provided build:
`/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ~/.scripts/hdr10plus_parser)`

My Ubuntu 18.04 is running GLIBC_2.27 which is the max available for 18.04. I looked online a bit and it seems you need to build the script with a lower version or?

**Comment by quietvoid** [comment]
> @quietvoid Getting this error now with your provided build:
> 
> ```
> /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ~/.scripts/hdr10plus_parser)
> ```
> 
> My Ubuntu 18.04 is running GLIBC_2.27 which is the max available for 18.04. I looked online a bit and it seems you need to build the script with a lower version or?

Yea that's the issue with glib linked linux builds, you'd have to build your own as they're not compatible across versions.

**Comment by FrostByteLXV** [comment]
Ahh, timely post for me.  I just got done encoding the movie 1917 and it errored 7 frames short of being complete 

Video encoding returned error exit code: -1073740940 (0xC0000374)

**Comment by theweasley** [comment]
@quietvoid okay that build worked. I've started an encode for 1917 also like the gentleman above me. This time there's no crash at the beginning so that's definitely been fixed. Now let's see if it returns any missing frames at the end.

**Comment by quietvoid** [comment]
> P.S: The fix that you've implemented, does it potentially impact other titles that used to work before with your previous release?

All the tests pass, so it shouldn't affect parsing.
It's just changing the output JSON to avoid mixing the different profiles, as that's what seems to cause problems with x265.
I don't really know if it will affect playback tonemapping either.

**Comment by quietvoid** [comment]
I've created an issue at x265: https://bitbucket.org/multicoreware/x265_git/issues/574/x265-segmentation-fault-when-using-hdr10
So for now I will just add a warning when this situation happens, with an optional flag that will "fix' it.

**Comment by theweasley** [comment]
Thanks for adding the optional flag in there to make it more graceful. Thanks for all your hard work and this excellent script. Hoping for a fix on the x265 end.

**Comment by FrostByteLXV** [comment]
Thanks both of you.  I'll give 1917 another go now.

**Comment by quietvoid** [comment]
Fixed in new release: https://github.com/quietvoid/hdr10plus_parser/releases/tag/0.2.9

**Comment by quietvoid** [comment]
It turns out there was actually an incorrect parsing, at least for Back to the Future.
However it only seems to have affected around ~10 frames.
This has been fixed in 0.3.0.
